### PR TITLE
Consolidate filter plugin

### PR DESCRIPTION
**Pull request by KB-perByte**
##### SUMMARY

Consolidate filter plugin 
This plugin presents collective structured data including all supplied facts grouping on common attributes mentioned.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

ansible.utils.consolidate
##### ADDITIONAL INFORMATION

```paste below

# Consolidated facts example
# ------------

##facts.yml
interfaces:
  - name: GigabitEthernet0/0
    enabled: true
    duplex: auto
    speed: auto
    note:
      - Connected green wire
  - name: GigabitEthernet0/1
    description: Configured by Ansible - Interface 1
    mtu: 1500
    speed: auto
    duplex: auto
    enabled: true
    note:
      - Connected blue wire
      - Configured by Paul
    vifs:
      - vlan_id: 100
        description: Eth1 - VIF 100
        mtu: 400
        enabled: true
        comment: Needs reconfiguration
      - vlan_id: 101
        description: Eth1 - VIF 101
        enabled: true
  - name: GigabitEthernet0/2
    description: Configured by Ansible - Interface 2 (ADMIN DOWN)
    mtu: 600
    enabled: false
l2_interfaces:
  - name: GigabitEthernet0/0
  - mode: access
    name: GigabitEthernet0/1
    trunk:
      allowed_vlans:
        - "11"
        - "12"
        - "59"
        - "67"
        - "75"
        - "77"
        - "81"
        - "100"
        - 400-408
        - 411-413
        - "415"
        - "418"
        - "982"
        - "986"
        - "988"
        - "993"
  - mode: trunk
    name: GigabitEthernet0/2
    trunk:
      allowed_vlans:
        - "11"
        - "12"
        - "59"
        - "67"
        - "75"
        - "77"
        - "81"
        - "100"
        - 400-408
        - 411-413
        - "415"
        - "418"
        - "982"
        - "986"
        - "988"
        - "993"
      encapsulation: dot1q
l3_interfaces:
  - ipv4:
      - address: 192.168.0.2/24
    name: GigabitEthernet0/0
  - name: GigabitEthernet0/1
  - name: GigabitEthernet0/2
  - name: Loopback888
  - name: Loopback999

##Playbook
  vars_files:
    - "facts.yml"
  tasks:
    - name: Build the facts collection
      set_fact:
        data_source:
          - data: "{{ interfaces }}"
            match_key: name
            prefix: interfaces
          - data: "{{ l2_interfaces }}"
            match_key: name
            prefix: l2_interfaces
          - data: "{{ l3_interfaces }}"
            match_key: name
            prefix: l3_interfaces

    - name: Combine all the facts based on match_keys
      set_fact:
        combined: "{{ data_source|ansible.utils.consolidate(fail_missing_match_value=False) }}"

##Output
# ok: [localhost] => {
#     "ansible_facts": {
#         "data_source": [
#             {
#                 "data": [
#                     {
#                         "duplex": "auto",
#                         "enabled": true,
#                         "name": "GigabitEthernet0/0",
#                         "note": [
#                             "Connected green wire"
#                         ],
#                         "speed": "auto"
#                     },
#                     {
#                         "description": "Configured by Ansible - Interface 1",
#                         "duplex": "auto",
#                         "enabled": true,
#                         "mtu": 1500,
#                         "name": "GigabitEthernet0/1",
#                         "note": [
#                             "Connected blue wire",
#                             "Configured by Paul"
#                         ],
#                         "speed": "auto",
#                         "vifs": [
#                             {
#                                 "comment": "Needs reconfiguration",
#                                 "description": "Eth1 - VIF 100",
#                                 "enabled": true,
#                                 "mtu": 400,
#                                 "vlan_id": 100
#                             },
#                             {
#                                 "description": "Eth1 - VIF 101",
#                                 "enabled": true,
#                                 "vlan_id": 101
#                             }
#                         ]
#                     },
#                     {
#                         "description": "Configured by Ansible - Interface 2 (ADMIN DOWN)",
#                         "enabled": false,
#                         "mtu": 600,
#                         "name": "GigabitEthernet0/2"
#                     }
#                 ],
#                 "match_key": "name",
#                 "prefix": "interfaces"
#             },
#             {
#                 "data": [
#                     {
#                         "name": "GigabitEthernet0/0"
#                     },
#                     {
#                         "mode": "access",
#                         "name": "GigabitEthernet0/1",
#                         "trunk": {
#                             "allowed_vlans": [
#                                 "11",
#                                 "12",
#                                 "59",
#                                 "67",
#                                 "75",
#                                 "77",
#                                 "81",
#                                 "100",
#                                 "400-408",
#                                 "411-413",
#                                 "415",
#                                 "418",
#                                 "982",
#                                 "986",
#                                 "988",
#                                 "993"
#                             ]
#                         }
#                     },
#                     {
#                         "mode": "trunk",
#                         "name": "GigabitEthernet0/2",
#                         "trunk": {
#                             "allowed_vlans": [
#                                 "11",
#                                 "12",
#                                 "59",
#                                 "67",
#                                 "75",
#                                 "77",
#                                 "81",
#                                 "100",
#                                 "400-408",
#                                 "411-413",
#                                 "415",
#                                 "418",
#                                 "982",
#                                 "986",
#                                 "988",
#                                 "993"
#                             ],
#                             "encapsulation": "dot1q"
#                         }
#                     }
#                 ],
#                 "match_key": "name",
#                 "prefix": "l2_interfaces"
#             },
#             {
#                 "data": [
#                     {
#                         "ipv4": [
#                             {
#                                 "address": "192.168.0.2/24"
#                             }
#                         ],
#                         "name": "GigabitEthernet0/0"
#                     },
#                     {
#                         "name": "GigabitEthernet0/1"
#                     },
#                     {
#                         "name": "GigabitEthernet0/2"
#                     },
#                     {
#                         "name": "Loopback888"
#                     },
#                     {
#                         "name": "Loopback999"
#                     }
#                 ],
#                 "match_key": "name",
#                 "prefix": "l3_interfaces"
#             }
#         ]
#     },
#     "changed": false
# }
# Read vars_file 'facts.yml'

# TASK [Combine all the facts based on match_keys] ****************************************************************************************************************
# ok: [localhost] => {
#     "ansible_facts": {
#         "combined": {
#             "GigabitEthernet0/0": {
#                 "interfaces": {
#                     "duplex": "auto",
#                     "enabled": true,
#                     "name": "GigabitEthernet0/0",
#                     "note": [
#                         "Connected green wire"
#                     ],
#                     "speed": "auto"
#                 },
#                 "l2_interfaces": {
#                     "name": "GigabitEthernet0/0"
#                 },
#                 "l3_interfaces": {
#                     "ipv4": [
#                         {
#                             "address": "192.168.0.2/24"
#                         }
#                     ],
#                     "name": "GigabitEthernet0/0"
#                 }
#             },
#             "GigabitEthernet0/1": {
#                 "interfaces": {
#                     "description": "Configured by Ansible - Interface 1",
#                     "duplex": "auto",
#                     "enabled": true,
#                     "mtu": 1500,
#                     "name": "GigabitEthernet0/1",
#                     "note": [
#                         "Connected blue wire",
#                         "Configured by Paul"
#                     ],
#                     "speed": "auto",
#                     "vifs": [
#                         {
#                             "comment": "Needs reconfiguration",
#                             "description": "Eth1 - VIF 100",
#                             "enabled": true,
#                             "mtu": 400,
#                             "vlan_id": 100
#                         },
#                         {
#                             "description": "Eth1 - VIF 101",
#                             "enabled": true,
#                             "vlan_id": 101
#                         }
#                     ]
#                 },
#                 "l2_interfaces": {
#                     "mode": "access",
#                     "name": "GigabitEthernet0/1",
#                     "trunk": {
#                         "allowed_vlans": [
#                             "11",
#                             "12",
#                             "59",
#                             "67",
#                             "75",
#                             "77",
#                             "81",
#                             "100",
#                             "400-408",
#                             "411-413",
#                             "415",
#                             "418",
#                             "982",
#                             "986",
#                             "988",
#                             "993"
#                         ]
#                     }
#                 },
#                 "l3_interfaces": {
#                     "name": "GigabitEthernet0/1"
#                 }
#             },
#             "GigabitEthernet0/2": {
#                 "interfaces": {
#                     "description": "Configured by Ansible - Interface 2 (ADMIN DOWN)",
#                     "enabled": false,
#                     "mtu": 600,
#                     "name": "GigabitEthernet0/2"
#                 },
#                 "l2_interfaces": {
#                     "mode": "trunk",
#                     "name": "GigabitEthernet0/2",
#                     "trunk": {
#                         "allowed_vlans": [
#                             "11",
#                             "12",
#                             "59",
#                             "67",
#                             "75",
#                             "77",
#                             "81",
#                             "100",
#                             "400-408",
#                             "411-413",
#                             "415",
#                             "418",
#                             "982",
#                             "986",
#                             "988",
#                             "993"
#                         ],
#                         "encapsulation": "dot1q"
#                     }
#                 },
#                 "l3_interfaces": {
#                     "name": "GigabitEthernet0/2"
#                 }
#             },
#             "Loopback888": {
#                 "interfaces": {},
#                 "l2_interfaces": {},
#                 "l3_interfaces": {
#                     "name": "Loopback888"
#                 }
#             },
#             "Loopback999": {
#                 "interfaces": {},
#                 "l2_interfaces": {},
#                 "l3_interfaces": {
#                     "name": "Loopback999"
#                 }
#             }
#         }
#     },
#     "changed": false
# }

```

### CODE REVIEW

File: plugins/filter/consolidate.py [hunk 1]
```diff
@@ -0,0 +1,1245 @@
...
+The consolidate filter plugin
+"""
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+DOCUMENTATION = """
+    name: consolidate
+    author: Sagar Paul (@KB-perByte)
+    version_added: "2.5.2"
```
Comment (ashwini-mhatre) [inline plugins/filter/consolidate.py:18]: this will be 2.6.0

File: plugins/plugin_utils/consolidate.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    Args:
+        data_sources (list): list of dicts as data sources
+        fail_missing_match_key (bool): Fails if match_keys not present in data set
+        fail_duplicate (bool): Fails if duplicate data present in a data
+    Returns:
+        list: list of unique keys based on specified match_keys
+    """
+    results, errors_match_key, errors_duplicate = [], [], []
+    for ds_idx, data_source in enumerate(data_sources):
+        match_key = data_source["match_key"]
```
Comment (ashwini-mhatre) [inline plugins/plugin_utils/consolidate.py:79]: does data_source always has this key?

Comment (KB-perByte) [reply plugins/plugin_utils/consolidate.py:79]: right I must add a `required: True` to them, Thanks

**Review by ashwini-mhatre — APPROVED** [review]

**Comment by cidrblock** [comment]
I've moved the black upgrade to #158 so this PR should have much less in it to review

File: plugins/filter/consolidate.py [hunk 3]
```diff
@@ -0,0 +1,1249 @@
...
+        type: bool
+      fail_missing_match_value:
+        description: Fail if a keys to match in not same accross all data sets.
+        type: bool
+      fail_duplicate:
+        description: Fail if duplicate values for any key is found.
+        type: bool
+"""
+
+EXAMPLES = r"""
```
Comment (cidrblock) [inline plugins/filter/consolidate.py:56]: Can we start with a super simple example?  maybe with a small a b c data set that shows how it works before showing a large data set? if you look at get_path or update_fact, they go from simple to complex in the examples.  We could end with an example using resource modules to get the data, or just the facts module to get different resources

Comment (cidrblock) [reply plugins/filter/consolidate.py:56]: Showing the behavior of the different failure option might e easier with a small sample data set as well

**Review by cidrblock — CHANGES_REQUESTED** [review]

File: plugins/filter/consolidate.py [hunk 4]
```diff
@@ -0,0 +1,1249 @@
...
+        suboptions:
+          data:
+            description: Specify facts data that gets consolidated.
+            type: raw
+            required: True
+          match_key:
+            description: Specify key to match on.
+            type: str
+            required: True
+          prefix:
```
[the remark this answers is not in the document]
Comment (KB-perByte) [reply plugins/filter/consolidate.py:41]: I was thinking `section_name` it's relatable!

File: plugins/plugin_utils/consolidate.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    Args:
+        matched_keys (list): list of unique keys based on specified match_keys
+        fail_missing_match_value (bool): Fail if match_key value is missing in a data set
+    Returns:
+        set: set of unique values
+    """
+    errors_match_values = []
+    all_values = set(itertools.chain.from_iterable(matched_keys))
+    if fail_missing_match_value:
+        for ds_idx, ds_values in enumerate(matched_keys):
```
Comment (cidrblock) [inline plugins/plugin_utils/consolidate.py:120]: enumerate(iterable, start=1), here too, let's start at 1

**Review by cidrblock — CHANGES_REQUESTED** [review]

**Comment by cidrblock** [comment]
Some smaller examples here, will need to be updated when prefix becomes `name` https://gist.github.com/cidrblock/8fbd4082f839e37d623858e4e7c81236

File: plugins/plugin_utils/consolidate.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+    Args:
+        matched_keys (list): list of unique keys based on specified match_keys
+        fail_missing_match_value (bool): Fail if match_key value is missing in a data set
+    Returns:
+        set: set of unique values
+    """
+    errors_match_values = []
+    all_values = set(itertools.chain.from_iterable(matched_keys))
+    if fail_missing_match_value:
```
[the remark this answers is not in the document]
Comment (KB-perByte) [reply plugins/plugin_utils/consolidate.py:119]: @cidrblock we are returning _all_values_ for the next function to iterate on the set, so I want to keep the action in `check_missing_match_values`  and not create the set again later. Moving it to the top would restrict the reusability of the return statement. Am I missing something here?

File: plugins/plugin_utils/consolidate.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+        fail_duplicate (bool): Fails if duplicate data present in a data
+    Returns:
+        list: list of unique keys based on specified match_keys
+    """
+    results, errors_match_key, errors_duplicate = [], [], []
+    for ds_idx, data_source in enumerate(data_sources):
+        match_key = data_source["match_key"]
+        ds_values = []
+
+        for dd_idx, data_dict in enumerate(data_source["data"]):
```
[the remark this answers is not in the document]
Comment (KB-perByte) [reply plugins/plugin_utils/consolidate.py:82]: I gave this a long thought then settled for what we would enjoy :D let me change it

**Review by cidrblock — APPROVED** [review]

**Pull request closed without merging**